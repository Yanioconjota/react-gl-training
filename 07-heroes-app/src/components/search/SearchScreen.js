import { useMemo } from "react";
import { useNavigate, useLocation } from "react-router-dom";
import queryString from "query-string";
import { useForm } from "../../hooks/useForm";
import { getHeroesByName } from "../../selectors/getHeroesByName";
import HeroCard from "../heroes/HeroCard";

export const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();

  //desestructuramos { q } porque es el primer argumento que viene en la query y le asignamos un valor por defecto para que no rompa, q va a ser el valor de searInput que le pasamos a la query
  const { q = "" } = queryString.parse(location.search);

  //usamos un custom hook, handleInput change le pasa el valor al handleSearch
  const [formValues, handleInputChange] = useForm({
    searchText: q,
  });

  const { searchText } = formValues;

  const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    //console.log(searchText);
    //navegamos a esta misma pantalla pero pasándo como params el valor de search text
    navigate(`?q=${searchText}`);
  };

  return (
    <>
      <h1>Hero Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <div className="h4">Buscar</div>
          <hr />
          <form onSubmit={handleSearch}>
            <input
              type="text"
              className="form-control"
              placeholder="Hero name"
              autoComplete="off"
              name="searchText"
              value={searchText}
              onChange={handleInputChange}
            />
            <button
              type="submit"
              className="btn btn-primary btn-block w-100 mt-3"
            >
              Buscar
            </button>
          </form>
        </div>
        <div className="col-7 search-hero">
          <h4>Resultados</h4>
          <hr />
          {q === "" ? (
            <div className="alert alert-info animate__animated animate__fadeInRight">
              Busca un héroe
            </div>
          ) : (
            heroesFiltered.length === 0 && (
              <div className="alert alert-danger animate__animated animate__fadeInRight">
                No se encotraron heroes: {q}
              </div>
            )
          )}

          {
            <div className="row">
              {heroesFiltered.map((hero) => (
                <HeroCard
                  key={hero.id}
                  {...hero}
                  customClass="col-12 col-md-4 animate__animated animate__fadeInRight"
                />
              ))}
            </div>
          }
        </div>
      </div>
    </>
  );
};
