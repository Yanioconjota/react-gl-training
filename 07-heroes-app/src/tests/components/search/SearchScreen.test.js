import "@testing-library/jest-dom";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

//Mock de la llamada al hook de react useNavigate --> IMPORTANTE: siempre empezar con la palabra 'mock'

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe('Pruebas con SearchScreen', () => {
  test('debe mostrarse correctamente con valores por defecto', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Busca un héroe');
  });

  test('Debe mostrar a Batman y el input con el valor del queryString', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchScreen />
      </MemoryRouter>
    );
    expect(wrapper.find('input').prop('value')).toBe('batman');
    expect(wrapper).toMatchSnapshot();
  });

  test('Debe mostrar un error si no se encuentra el héroe', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=bacman"]}>
        <SearchScreen />
      </MemoryRouter>
    );
    expect(wrapper.find('.alert-danger').text().trim()).toBe('No se encotraron heroes: bacman');
    expect(wrapper).toMatchSnapshot();
  });

  test('Debe llamar el navigate', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchScreen />
      </MemoryRouter>
    );
    
    //Se simula el change en el input
    wrapper.find('input').simulate('change', {
      target: {
        name: 'searchText',
        value: 'batman'
      }
    });

    // Se simula el evento submit para ello debemos mockear la llamada al hook useNavigate --> Verificar antes del describe
    wrapper.find('form').prop('onSubmit')({
      preventDefault(){}
    });

    expect(mockNavigate).toHaveBeenCalled();
    expect(mockNavigate).toHaveBeenCalledWith('?q=batman');

  });
});