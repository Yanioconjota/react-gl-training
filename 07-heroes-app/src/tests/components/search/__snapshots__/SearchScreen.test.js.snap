// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con SearchScreen Debe mostrar a Batman y el input con el valor del queryString 1`] = `
Array [
  <h1>
    Hero Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <div
        className="h4"
      >
        Buscar
      </div>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Hero name"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-primary btn-block w-100 mt-3"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7 search-hero"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="row"
      >
        <div
          className="card col-12 col-md-4 animate__animated animate__fadeInRight"
        >
          <div
            className="row no-gutters"
          >
            <img
              alt="Batman"
              className="card-img-top"
              src="./assets/img/dc-batman.jpg"
            />
            <div
              className="card-body px-4"
            >
              <h2
                className="card-title"
              >
                Batman
              </h2>
              <p>
                DC Comics
              </p>
              <a
                className="btn w-100 mt-3 btn-info"
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Mas info...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas con SearchScreen Debe mostrar un error si no se encuentra el héroe 1`] = `
Array [
  <h1>
    Hero Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <div
        className="h4"
      >
        Buscar
      </div>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Hero name"
          type="text"
          value="bacman"
        />
        <button
          className="btn btn-primary btn-block w-100 mt-3"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7 search-hero"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-danger animate__animated animate__fadeInRight"
      >
        No se encotraron heroes: 
        bacman
      </div>
      <div
        className="row"
      />
    </div>
  </div>,
]
`;

exports[`Pruebas con SearchScreen debe mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Hero Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <div
        className="h4"
      >
        Buscar
      </div>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Hero name"
          type="text"
          value=""
        />
        <button
          className="btn btn-primary btn-block w-100 mt-3"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7 search-hero"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info animate__animated animate__fadeInRight"
      >
        Busca un héroe
      </div>
      <div
        className="row"
      />
    </div>
  </div>,
]
`;
